Slack client features

pcn opened the pull request:
### What does this PR do?

This enhances the slack engine

### What issues does this PR fix or reference?

None.

### New Behavior

This slack engine will do the following:

1. Configuration is based around having groups.  Each group is a collection of commands, aliases, and target specifications that will restrict commands to just the slack users in that groupu
2. Configuration via the saltmaster's pillars so that users/groups/commands/aliases/targets can be re-configured without updating the engine config
3. You can fire to the event bus as well as running commands by configuration.
4. if there is a problem initially connecting to the slack api, retries are attempted with some backoffs so that the connection doesn't get throttled by slack which was happening to me while testing (via the engine dying and coming back quickly and failing again). 
5. When firing to the event bus, the user name and channel name are recorded and added to the data on the event bus so that responding to events can use this info for decision making.

### Tests written?

No

 but I'm open to suggestions that would help here.


Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

pcn committed:
Break out group settings to a separate function

- Move fire() outside of main()
- Updated docs to only have group-based config
- This reads permissions on a per-group basis
, Add a permissions checker

pcn committed:
Move the message acquistion to a generator

- Move more of the logic out of start()
- w/o control set, don't send messages to event bus
- If control is not set, don't fire non-text events
  to the salt event bus by default.
- Fix tgt_type -> expr_form; it looks like the tgt_type arg has been
  depricated in favor of expr_form.
- Also, for troubleshooting actually write to the file slack should send
  for now.

pcn committed:
make firing to the event bus a config option

- Bring most log messages down to debug level
- Fix missing tag to the command runner
- Remove leading _ from msg

pcn committed:
Firing to the event bus is broken

pcn committed:
Groups from pillars

- Get groups from a pillar
- Pillar group data is the same syntax as the config file data
- In each group, you can add default targeting for that group or each command
- Fix function names and None tgt_type
- Note about the async future of this

pcn committed:
Make handling of args and kwargs more DRY

pcn committed:
We get data from pillars

pcn committed:
Make running commands async and fix events

- Fix lots of lint warnings and one typo
- Fix a typo

pcn committed:
Fix runner invocation and better error msgs

- Add args to the low for runner.async
- Call out when types are wrong in group config

pcn committed:
Make fire_all more useful

- Include user and channel info

Removed some dead code too

salt-jenkins commented:
@pcn, thanks for your PR! By analyzing the history of the files in this pull request, we identified @garethgreenaway, @jacobhammons and @cachedout to be potential reviewers.

cachedout commented:
Hi @pcn 

Thanks for this!

Could you please fix the lint errors before we review this? Thanks.

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/11126/violations/

pcn committed:
Get rid of obvious pylint warnings

pcn committed:
Init logging before its first invocation

pcn committed:
The tests don't work as-is

pcn committed:
Fix a couple of more pylint issues

pcn commented:
I don't know what to do about the `missing-whitespace-after-comma` E231 errors.  There are whitespaces there.

Commit:
pylint disable

pcn committed:
Disable checks that are bogus or desired

- bogus: missing-whitespace-after-comma... there are so commas!
- desired: broad exception catching on the basis that I don't know
  what exceptions may happen here yet.

pcn committed:
Merge branch 'slack-client-features' of github.com:pcn/salt into slack-client-features

pcn committed:
This was in the wrong place

cachedout commented:
This is really nicely done. Thanks a ton, @pcn
